### ISSUE TITLE
flow-build-info.json isn't found in production build uberjar

### ISSUE THREAD

**Issue by mvysny** [vaadin/flow#7060]
Related to https://github.com/vaadin/flow/issues/6965

The production uberjar fails with

```
Caused by: java.lang.IllegalStateException: Unable to determine mode of operation. To use npm mode, ensure 'flow-build-info.json' exists on the classpath. With Maven, this is handled by the 'prepare-frontend' goal. To use compatibility mode, add the 'flow-server-compatibility-mode' dependency. If using Vaadin with Spring Boot, instead set the property 'vaadin.compatibilityMode' to 'true' in 'application.properties'.
	at com.vaadin.flow.server.DeploymentConfigurationFactory.verifyMode(DeploymentConfigurationFactory.java:382)
	at com.vaadin.flow.server.DeploymentConfigurationFactory.lambda$readBuildInfo$1ac205d3$1(DeploymentConfigurationFactory.java:230)
	at com.vaadin.flow.server.DefaultDeploymentConfiguration.checkCompatibilityMode(DefaultDeploymentConfiguration.java:275)
	at com.vaadin.flow.server.DefaultDeploymentConfiguration.<init>(DefaultDeploymentConfiguration.java:120)
```

I can reproduce this issue easily with my https://github.com/mvysny/vaadin14-embedded-jetty project:

* By default it uses 14.0.12 and it works: build the project with `mvn -C clean package -Pproduction` then go to `target/` and run `java -jar vaadin14-embedded-jetty-1.0-SNAPSHOT-uberjar.jar` and browse to localhost:8080 -  everything works.
* However, edit the pom.xml and set vaadin version to 14.0.13 or 14.0.14, build the project with `mvn -C clean package -Pproduction` then go to `target/` and run `java -jar vaadin14-embedded-jetty-1.0-SNAPSHOT-uberjar.jar` and browse to localhost:8080 - it will crash.

The `META-INF/VAADIN/config/flow-build-info.json` exists in both jars, however it seems it's being ignored somehow in 14.0.13 and 14.0.14...

Sounds like 14.0.13+ only checks `fragment.xml` (which is for some reason ignored by Jetty in uberjar mode).

**Comment by mvysny** [comment]
Debugging flow-server-2.0.18 `getResourceFromClassloader:284, DeploymentConfigurationFactory`, the json is indeed found at `jar:file:/home/mavi/work/my/vaadin14-embedded-jetty/target/vaadin14-embedded-jetty-1.0-SNAPSHOT-uberjar.jar!/META-INF/VAADIN/config/flow-build-info.json` at line 272, but it's filtered out at line 277.

Therefore some kind of fallback triggers, the `getPossibleJarResource()`, which founds the generated webpack at `jar:file:/home/mavi/work/my/vaadin14-embedded-jetty/target/vaadin14-embedded-jetty-1.0-SNAPSHOT-uberjar.jar!/webpack.generated.js`. However, `countInstances()` at 334 returns just `1` and therefore the whole thing returns null and reports that there is no json on classpath.

I don't understand the logic behind that function, no way I can fix that.

**Comment by mvysny** [comment]
From `DefaultDeploymentConfiguration:272` i judge that if the compat mode is explicity set (`explicitlySet` is true), then Vaadin isn't bothered by "missing" json, and truly calling

```
System.setProperty("vaadin.compatibilityMode", "false");
```
helps.

**Comment by mvysny** [comment]
Workaround is to downgrade to Vaadin 14.0.12, or downgrade flow-server to 2.0.17.

**Comment by caalador** [comment]
The problem here is that what we are working against is to not get the build-info for an add-on that is packaged wrong as this has led to some really strange errors where we try to use add-on dev system paths (which will ofc fail)

Now in 14.0.9 we filtered out any info inside a jar, in 14.0.11 we got a simple filter that accepted a build info inside a jar if productionMode was set and the build info contained prodmode=true as this will not use file system paths. in 14.0.13 the issue was fixed so that we look at the file webpack.generated.js which resides in flow-server.jar so in the case where we run a jar this should be in the library file which would make it have the path exec.jar/lib/flow-server.jar/webpack.generated.js where we would know that we are running a jar container as the file is inside 2 jars (exec and lib jar) and we can take thebuild info inside one jar.

Now in your uberjar build you don't have lib jars and they are flattened meaning our check for the jar in a jar (webpack.generated.js) says that it's inside one jar so we can't accept any jar build info as that would form our perspective be an add-on jar and probably wrong.

**Comment by mvysny** [comment]
I think I got it, thank you very much for a good explanation :) However I'm not sure what kind of situation the 14.0.13 fix is trying to prevent: if we're trying to prevent loading add-on build-info.json, the code in 14.0.13 won't prevent that since webpack.generated.js will come from flow-server.jar and the build-info.json will come from the addon jar?

Regardless. In light of this, I wonder whether we want to have a support for flattened uberjars. They're a bit peculiar (e.g. you can't repeat certain resources e.g. in META-INF/services).

Also, there is probably a deeper issue here. I have a zip with jars located in `libs/` folder; then I run `java -cp all-those-jars Main` and that one also fails with can't find `flow-build-info.json`. And that one has definitely webpack.generated.js located in flow-server.jar and json in app.jar...

**Comment by caalador** [comment]
No in the case we have now when the generated.js is in flow-server.jar it is expected that it is inside the application jar so then it will have 2 jar! in the path and in that case we can select a build-info that has one jar! in the path as any add-on would have 2 jar!

For the latter case I would expect that because the jars are on the classpath directly the generated.js in the flow-server.jar will show with a path containing only one jar!

This is a really complex case as we don't know if we are running as a war or a jar and a flattened uberjar brings in new problems. What if you have a flattened uberjar with an add-on that has flow-build-info.json, will one of the files override the other or does the build throw?

**Comment by mvysny** [comment]
> No in the case we have now when the generated.js is in flow-server.jar it is expected that it is inside the application jar so then it will have 2 jar! in the path and in that case we can select a build-info that has one jar! in the path as any add-on would have 2 jar!

Understood, that's the jar-in-a-jar case.

> For the latter case I would expect that because the jars are on the classpath directly the generated.js in the flow-server.jar will show with a path containing only one jar!

Precisely. And now I understand why the zipped-jar doesn't work: it's not possible to tell a legitimate json apart from a json accidentally packaged in by add-on.

> This is a really complex case as we don't know if we are running as a war or a jar and a flattened uberjar brings in new problems. What if you have a flattened uberjar with an add-on that has flow-build-info.json, will one of the files override the other or does the build throw?

Exactly - that's the limitation of uberjar and it's not even clear which one wins. Uberjar is definitely tricky.

I was going to suggest to optionally disable this new check which came with 14.0.13, but I can now see how that would allow you to shoot in a foot.

Please let me think on this.

**Comment by mvysny** [comment]
To support this flawed flat uberjar case, would you please considering adding a config option which disables the 2 `jar!` detection added in Vaadin 14.0.13 please? If somebody wants to use flat uberjar then he should understand inherent limitations coming with that technology.

**Comment by TomaszPadzik** [comment]
Unfortunately, this verification prevents deployment on weblogic server with production mode set to true in flow-build-info.json. When deploying app on weblogic server as exploded archive, Weblogic put all classes from war, included flow-build-info.json, into jar. New filter in getResourceFromClassloader() founds flow-build-info.json in jar build by weblogic, and then tries to found more instances of jar, which of course fails. This check makes impossible to read flow-build-info.json on weblogic server, please consider disabling this check.

**Comment by mvysny** [comment]
Awesome work, thank you guys so much! Will the fix appear in Vaadin 14.1 please?

**Comment by caalador** [comment]
It should come with 14.1.1 that by what I remember should follow quite soon after 14.1.0

**Comment by mvysny** [comment]
Thank you, I verify that the fix works properly in 14.1.2
